### ISSUE TITLE
Training instability and non-convergence with TAPIR on kubric-e dataset across multiple GPUs

### ISSUE THREAD

**Issue by swarnim-j** [google-deepmind/tapnet#130]
I'm experiencing issues training the TAPIR model on the kubric-e dataset across different GPU configurations with PyTorch lightning (both checkpoint and non-checkpoint model). The losses either don't converge or show high instability, even with gradient clipping.

**Dataset**: kubric-e

**Observed Issues**
_Loss Instability:_
When training from scratch: Losses fluctuate between 5 and 100+
With checkpointed model: Losses fluctuate between 0.5 and 10+

_Different GPU/Batch Size Configurations Tested:_
2 GPUs with batch size 2 (2 x 24)
1 GPU with batch size 4 (1 x 48)
1 GPU with batch size 2 (1 x 48)

train_tapir.py:
```
if __name__ == '__main__':
    checkpoint_state_dict = torch.load("/scratch/shared/beegfs/shivanim/tapir/tapir_checkpoint_panning.pt")
    model = tapir_model.TAPIR()
    model.load_state_dict(checkpoint_state_dict, strict=False)
    #model = model.to(torch.device('cuda'))
    data_root = os.path.join("", "/scratch/shared/beegfs/shivanim/tapir/tapvid-kinetics/pickle_folder/")
    train_dataset = TapVidDataset(dataset_type="kinetics", data_root=data_root)
    train_loader = DataLoader(
         train_dataset,
         batch_size=1,
         shuffle=True,
         num_workers=0,
         pin_memory=True,
         collate_fn=collate_fn,
         drop_last=True,
    )
    trainer = Trainer(accelerator='gpu')
    trainer.fit(model, train_loader)
    from torch.optim import Adam
    optimizer = Adam(model().parameters(), lr=1e-3)
```
tapir_model.py:
```
def training_step(self, batch, batch_idx):
    if batch.video == []:
      return None
    frames = batch.video
    #TODO: ask what is this getting?
    query_points = batch.query_points
    gt_occluded = batch.occluded.to(float)#.cuda()
    gt_target_points = batch.target_points
    shape = frames.shape
    outputs = self.forward(frames, query_points)
    if outputs is None:
      return None 
    tracks, occlusions, expected_dist = outputs['tracks'][0], outputs['occlusion'][0], outputs['expected_dist'][0]
    loss_huber, loss_occ, loss_prob = model_utils.tapnet_loss(tracks, occlusions, gt_target_points, gt_occluded, shape)
    # print("loss_huber", loss_huber)
    # print("loss_occ", loss_occ)
    # print("loss_prob", loss_prob)
    visibles = self.postprocess_occlusions(occlusions, expected_dist).detach().cpu().numpy()
    tracks = tracks.detach().cpu().numpy()
    loss = loss_huber + loss_occ + loss_prob
    
    # Print losses every 50 steps
    if batch_idx % 50 == 0:
        print(f"Batch {batch_idx} - Huber Loss: {loss_huber:.4f}, Occlusion Loss: {loss_occ:.4f}, Prob Loss: {loss_prob:.4f}, Total Loss: {loss:.4f}")
    
    return loss
    #return {"tracks":tracks, "visibles":visibles, "loss":loss_huber, "loss_occ":loss_occ, "loss_prob":loss_prob} 
```
tap_vid_datasets.py (adapted from  @co-tracker):
```
# Copyright (c) Meta Platforms, Inc. and affiliates.
# All rights reserved.

# This source code is licensed under the license found in the
# LICENSE file in the root directory of this source tree.

import os
import io
import glob
import torch
import pickle
import numpy as np
import mediapy as media
from os import path
from tqdm import tqdm
import tensorflow as tf
from PIL import Image
from typing import Mapping, Tuple, Union

from typing import Any, Optional
from dataclasses import dataclass
#from cotracker.datasets.utils import TapirData

DatasetElement = Mapping[str, Mapping[str, Union[np.ndarray, str]]]

@dataclass(eq=False)
class TAPIRData:
    """
    Dataclass for storing video tracks data.
    """

    video: torch.Tensor  # B, S, C, H, W
    target_points: torch.Tensor  # B, S, N, 2
    occluded: torch.Tensor  # B, S, N
    # optional data
    log_frames: Optional[torch.Tensor] = None
    valid: Optional[torch.Tensor] = None  # B, S, N
    segmentation: Optional[torch.Tensor] = None  # B, S, 1, H, W
    seq_name: Optional[str] = None
    query_points: Optional[torch.Tensor] = None  # TapVID evaluation format

import torch.nn.functional as F

def collate_fn(batch):
    """
    Collate function for video tracks data with padding.
    """

    min_queries = min(b.target_points.shape[1] for b in batch if b.target_points is not None)
    min_videos = min(b.video.shape[1] for b in batch if b.video is not None)

    video = torch.stack([b.video[:, :min_videos, ...] for b in batch if b.video is not None], dim=0)[0]
    log_frames = torch.stack([b.log_frames[:min_videos, ...] for b in batch if b.log_frames is not None], dim=0)[0]
    target_points = torch.stack([b.target_points[:, :min_queries, :min_videos, ...] for b in batch if b.target_points is not None], dim=0)[0]
    occluded = torch.stack([b.occluded[:, :min_queries, :min_videos] for b in batch if b.occluded is not None], dim=0)[0]
    query_points = segmentation = None
    
    if any(b.query_points is not None for b in batch):
        query_points = torch.stack([b.query_points[:, :min_queries, ...] for b in batch if b.query_points is not None], dim=0)[0]
    if any(b.segmentation is not None for b in batch):
        segmentation = torch.stack([b.segmentation[:, :min_queries, ...] for b in batch if b.segmentation is not None], dim=0)[0]
    
    seq_name = [b.seq_name for b in batch]

    return TAPIRData(
        video=video,
        target_points=target_points,
        occluded=occluded,
        segmentation=segmentation,
        seq_name=seq_name,
        query_points=query_points,
        log_frames=log_frames,
    )

def resize_video(video: np.ndarray, output_size: Tuple[int, int]) -> np.ndarray:
    """Resize a video to output_size."""
    # If you have a GPU, consider replacing this with a GPU-enabled resize op,
    # such as a jitted jax.image.resize.  It will make things faster.
    return media.resize_video(video, output_size)

def sample_queries_first(
    target_occluded: np.ndarray,
    target_points: np.ndarray,
    frames: np.ndarray,
) -> Mapping[str, np.ndarray]:
    """Package a set of frames and tracks for use in TAPNet evaluations.
    Given a set of frames and tracks with no query points, use the first
    visible point in each track as the query.
    Args:
      target_occluded: Boolean occlusion flag, of shape [n_tracks, n_frames],
        where True indicates occluded.
      target_points: Position, of shape [n_tracks, n_frames, 2], where each point
        is [x,y] scaled between 0 and 1.
      frames: Video tensor, of shape [n_frames, height, width, 3].  Scaled between
        -1 and 1.
    Returns:
      A dict with the keys:
        video: Video tensor of shape [1, n_frames, height, width, 3]
        query_points: Query points of shape [1, n_queries, 3] where
          each point is [t, y, x] scaled to the range [-1, 1]
        target_points: Target points of shape [1, n_queries, n_frames, 2] where
          each point is [x, y] scaled to the range [-1, 1]
    """
    valid = np.sum(~target_occluded, axis=1) > 0
    target_points = target_points[valid, :]
    target_occluded = target_occluded[valid, :]

    query_points = []
    for i in range(target_points.shape[0]):
        index = np.where(target_occluded[i] == 0)[0][0]
        x, y = target_points[i, index, 0], target_points[i, index, 1]
        query_points.append(np.array([index, y, x]))  # [t, y, x]

    if len(query_points) < 2:
        import pdb;pdb.set_trace()
        print("from tap_vid_datasets, query less than size 2...")

    query_points = np.stack(query_points, axis=0)

    return {
        "video": frames[np.newaxis, ...],
        "query_points": query_points[np.newaxis, ...],
        "target_points": target_points[np.newaxis, ...],
        "occluded": target_occluded[np.newaxis, ...],
    }


def sample_queries_strided(
    target_occluded: np.ndarray,
    target_points: np.ndarray,
    frames: np.ndarray,
    query_stride: int = 5,
) -> Mapping[str, np.ndarray]:
    """Package a set of frames and tracks for use in TAPNet evaluations.

    Given a set of frames and tracks with no query points, sample queries
    strided every query_stride frames, ignoring points that are not visible
    at the selected frames.

    Args:
      target_occluded: Boolean occlusion flag, of shape [n_tracks, n_frames],

        where True indicates occluded.
      target_points: Position, of shape [n_tracks, n_frames, 2], where each point
        is [x,y] scaled between 0 and 1.
      frames: Video tensor, of shape [n_frames, height, width, 3].  Scaled between
        -1 and 1.
      query_stride: When sampling query points, search for un-occluded points
        every query_stride frames and convert each one into a query.

    Returns:
      A dict with the keys:
        video: Video tensor of shape [1, n_frames, height, width, 3].  The video
          has floats scaled to the range [-1, 1].
        query_points: Query points of shape [1, n_queries, 3] where
          each point is [t, y, x] scaled to the range [-1, 1].
        target_points: Target points of shape [1, n_queries, n_frames, 2] where
          each point is [x, y] scaled to the range [-1, 1].
        trackgroup: Index of the original track that each query point was
          sampled from.  This is useful for visualization.
    """
    tracks = []
    occs = []
    queries = []
    trackgroups = []
    total = 0
    trackgroup = np.arange(target_occluded.shape[0])
    for i in range(0, target_occluded.shape[1], query_stride):
        mask = target_occluded[:, i] == 0
        query = np.stack(
            [
                i * np.ones(target_occluded.shape[0:1]),
                target_points[:, i, 1],
                target_points[:, i, 0],
            ],
            axis=-1,
        )
        queries.append(query[mask])
        if len(query[mask]) < 2:
            print("from tap_vid_datasets, query less than size 2...")
        tracks.append(target_points[mask])
        occs.append(target_occluded[mask])
        if len(occs) < 1:
            print("from tap_vid_datasets, occs less than size 1...")

        trackgroups.append(trackgroup[mask])
        total += np.array(np.sum(target_occluded[:, i] == 0))

    if len(occs) < 1:
         print("from tap_vid_datasets, outside for loop, occs less than size 1...")

    return {
        "video": frames[np.newaxis, ...],
        "query_points": np.concatenate(queries, axis=0)[np.newaxis, ...],
        "target_points": np.concatenate(tracks, axis=0)[np.newaxis, ...],
        "occluded": np.concatenate(occs, axis=0)[np.newaxis, ...],
        "trackgroup": np.concatenate(trackgroups, axis=0)[np.newaxis, ...],
    }


class TapVidDataset(torch.utils.data.Dataset):
    def __init__(
        self,
        data_root,
        dataset_type="kubric",
        resize_to_256=True,
        queried_first=True,
    ):
        self.dataset_type = dataset_type
        self.resize_to_256 = resize_to_256
        self.queried_first = queried_first
        self.all_paths = []
        
        if self.dataset_type == "kubric":
            # Just store the paths, don't load the data
            self.data_paths = sorted(glob.glob(os.path.join(data_root, "*.pkl")))
            # Create a mapping from index to (batch_file_idx, sequence_idx)
            self._create_index_mapping()
            print(f"Found {len(self.index_map)} unique videos in {data_root}")
            return
            
        elif self.dataset_type == "kinetics":
            self.all_paths = tf.io.gfile.glob(path.join(data_root, '*_of_0010.pkl'))
            points_dataset = []
            for pickle_path in self.all_paths:
                with open(pickle_path, "rb") as f:
                    data = pickle.load(f)
                    if isinstance(data, dict):
                        data = list(data.values())
                    points_dataset = points_dataset + data
            self.points_dataset = points_dataset
            
        elif self.dataset_type == "pst":
            data_root = "/scratch/shared/beegfs/shivanim/tapvid3d_drive/"
            self.points_dataset = tf.io.gfile.glob(path.join(data_root, '*.npz'))
            
        else:
            with open(data_root, "rb") as f:
                self.points_dataset = pickle.load(f)
            if self.dataset_type == "davis":
                self.video_names = list(self.points_dataset.keys())
        print(f"Found {len(self.points_dataset)} unique videos in {data_root}")

    def _create_index_mapping(self):
        """Create mapping from dataset index to (batch_file_idx, sequence_idx)"""
        self.index_map = []
        for batch_idx, batch_path in tqdm(enumerate(self.data_paths), total=len(self.data_paths), desc="Processing batches"):
            with open(batch_path, "rb") as f:
                batch_data = pickle.load(f)
                for seq_idx in range(len(batch_data)):
                    self.index_map.append((batch_idx, seq_idx))
    
    def __getitem__(self, index):
        if self.dataset_type == "kubric":
            # Get the batch file and sequence indices
            batch_idx, seq_idx = self.index_map[index]
            # Load only the required batch
            with open(self.data_paths[batch_idx], "rb") as f:
                batch_data = pickle.load(f)
            # Get the specific sequence
            sequence = batch_data[seq_idx]
            
            frames = sequence['video']
            target_occ = sequence['occluded']
            target_points = sequence['target_points']
            log_frames = sequence['log_frames']
            
            if self.queried_first:
                converted = sample_queries_first(target_occ, target_points, frames)
            else:
                converted = sample_queries_strided(target_occ, target_points, frames)
                
            converted['log_frames'] = log_frames
            
            return TAPIRData(
                torch.from_numpy(converted["video"]),
                torch.from_numpy(converted["target_points"]),
                torch.from_numpy(converted["occluded"]),
                seq_name=str(index),
                query_points=torch.from_numpy(converted["query_points"]),
                log_frames=torch.from_numpy(log_frames)
            )
            
        elif self.dataset_type == "kinetics":
            video_name = self.video_names[index]
            video = self.points_dataset[video_name]
            frames = video["video"]
            target_occ = self.points_dataset[video_name]["occluded"]
            target_points = self.points_dataset[video_name]["points"]

        if isinstance(frames[0], bytes):
            # TAP-Vid is stored and JPEG bytes rather than `np.ndarray`s.
            def decode(frame):
                byteio = io.BytesIO(frame)
                img = Image.open(byteio)
                return np.array(img)

            frames = np.array([decode(frame) for frame in frames])

        if self.resize_to_256:
            frames = resize_video(frames, [256, 256])
            target_points *= np.array([255, 255])  # 1 should be mapped to 256-1
        else:
            target_points *= np.array([frames.shape[2] - 1, frames.shape[1] - 1])

        log_frames = frames
        frames = frames.astype(np.float32) / 255.0 * 2.0 - 1.0

        valid = np.sum(~target_occ, axis=1) > 0
        if sum(valid) == 0:
            print("valid...", valid)
            return TAPIRData(
                [],
                [],
                [],
                seq_name=str(video_name),
                query_points=[],
                log_frames=[]
             )

        if self.queried_first and self.dataset_type != "pst":
            converted = sample_queries_first(target_occ, target_points, frames)
        elif ~self.queried_first and self.dataset_type != "pst":
            converted = sample_queries_strided(target_occ, target_points, frames)
        else:
            print("query points already computed")

        converted['log_frames'] = log_frames
        occluded = torch.from_numpy(converted["occluded"])
        log_frames = torch.from_numpy(log_frames)
        target_pts = torch.from_numpy(converted["target_points"])
        rgbs = torch.from_numpy(converted["video"])
        query_points = torch.from_numpy(converted["query_points"])

        return TAPIRData(
            rgbs,
            target_pts,
            occluded,
            seq_name=str(video_name),
            query_points=query_points,
            log_frames=log_frames
        )

    def __len__(self):
        if self.dataset_type == "kubric":
            return len(self.index_map)
        return len(self.points_dataset)
```

**Comment by yangyi02** [comment]
Hi, one quick try would be to add loss to all intermediate outputs as well. Since Tapir is an iterative refinement method, every single step output is supervised to get close to ground truth. Please see the paper for details. Thanks.
